server/src/main/java/org/elasticsearch/cluster/SafeClusterStateSupplier.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.cluster;
+
+import java.util.Optional;
+
+import static org.elasticsearch.gateway.GatewayService.STATE_NOT_RECOVERED_BLOCK;
+
+/**
+ * Utility to access {@link ClusterState} only when it is "ready", where "ready" means that we received a first clusterChanged event
+ * with no global block of type {@code STATE_NOT_RECOVERED_BLOCK}
+ * This guarantees that:
+ * - the initial cluster state has been set (see
+ * {@link org.elasticsearch.cluster.service.ClusterApplierService#setInitialState(ClusterState)});
+ * - the initial recovery process has completed.
+ */
+public class SafeClusterStateSupplier implements ClusterStateSupplier, ClusterStateListener {
+    private volatile ClusterState currentClusterState;
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (isInitialized() || event.state().blocks().hasGlobalBlock(STATE_NOT_RECOVERED_BLOCK) == false) {
+            currentClusterState = event.state();
```

Comment:
I'm not sure I quite grasp this if condition. Perhaps a comment would help?

It looks like once we've received the first event, we start ignoring `STATE_NOT_RECOVERED_BLOCK`. Why is that?

Reply:
That's the purpose. We just care about the _initial_ recovery; after that, we consider the cluster state as available.

Reply:
Ok. But can `STATE_NOT_RECOVERED_BLOCK` happen again later, after initialization?
- If so, probably this code could use a comment explaining why we want to keep accepting new cluster states in that case but not during initialization.
- If not, then we can get rid of the `isInitialized` call.

Reply:
Yes it can, but we care about the initial recovery yet. I'll add a comment about that.